### PR TITLE
feat: add jsonrpc prt tests

### LINKED ISSUE #682: Tests: JSON-RPC correctness, limit & security coverage, and OpenRPC schema compliance
### Context / Problem  
The JSON-RPC server is Rollups-Node’s public contract.  
There is **no automated test suite** that

1. Verifies **functional correctness** for every method under normal, edge-case, and error conditions.  
2. Exercises **boundary & security** behaviour (oversize inputs, malformed JSON, injection vectors, race conditions).  
3. Confirms responses match the shapes defined in `internal/jsonrpc/jsonrpc-discover.json`.

Lacking this coverage, regressions can ship unnoticed, breaking clients or exposing vulnerabilities.

---

### Suggested Solution  

1. **Spec-driven inventory**  
   * Parse the OpenRPC document at test start to obtain the full list of declared methods.  
   * A “contract-guard” test fails if any listed method lacks a dedicated test file, ensuring future coverage.

2. **Execution harness**  
   * **Integration layer** – spawn the standalone executable `./cartesi-rollups-jsonrpc-api` as a subprocess on a random local port for end-to-end tests.  
   * **Unit layer** – import the internal handler package and use `httptest.NewServer` for fast, handler-level tests that don’t require the full binary.  
   * Both layers point to a developer-supplied Postgres DSN (e.g., `PG_DSN_FOR_TESTS`) and load SQL fixtures that create:  
     * Minimal happy-path data  
     * Edge-case rows (max uint256, empty epochs, huge payloads)

3. **Table-driven cases per method**  

   | Bucket        | Coverage goals                                        |
   |---------------|-------------------------------------------------------|
   | **Happy**     | Valid inputs → deterministic, domain-correct result   |
   | **Boundary**  | Max/min values, pagination edges, numeric overflows   |
   | **Invalid**   | Missing params, wrong types, unknown IDs              |
   | **Security**  | Injection strings, extremely large payloads, replay   |
   | **Concurrency** | 50–100 goroutines issuing mixed requests (`go test -race`) |

   * Each response is first validated against its JSON Schema (shape).  
   * Additional assertions check **content correctness** (values, ordering, counts, error codes/messages).

4. **CI integration**  
   * GitHub Actions spins up a Postgres container, sets `PG_DSN_FOR_TESTS`, then runs `go test ./internal/jsonrpc/... -race`.  
   * Failures on schema drift **or** behavioural mismatch block the merge.

---

### Deliverables & File Layout  

| File / Dir                                     | Purpose                                                                          |
|------------------------------------------------|----------------------------------------------------------------------------------|
| `internal/jsonrpc/tests/contract_guard_test.go`| Parses `jsonrpc-discover.json`; fails if any method lacks its own test cases.     |
| `internal/jsonrpc/tests/method_<name>_test.go` | Table-driven cases for each RPC method, e.g. `method_getEpochs_test.go`.          |
| `internal/jsonrpc/tests/security_test.go`      | Cross-method security checks (injection, oversize payloads, DoS attempts).        |
| `internal/jsonrpc/tests/concurrency_test.go`   | High-RPS mixed-call stress under `go test -race`.                                 |
| `internal/jsonrpc/tests/testutil/`             | Helpers: subprocess launcher for `cartesi-rollups-jsonrpc-api`, fixture loader, JSON-Schema validator.|

---

### Acceptance Criteria  

| # | Scenario                                                             | Expected outcome                                                             |
|---|----------------------------------------------------------------------|------------------------------------------------------------------------------|
| 1 | Run `go test ./internal/jsonrpc/... -race` with local Postgres       | All tests pass; handler code coverage ≥ 80 %                                  |
| 2 | Intentionally change an RPC handler to return an incorrect value     | Corresponding method test fails with a clear assertion diff                  |
| 3 | Remove a field required by the OpenRPC spec                          | Shape validation test fails with descriptive error                           |
| 4 | Add a new method to the spec without adding tests                    | `contract_guard_test.go` fails, listing the missing method                   |
| 5 | Send payload `"\\"' OR 1=1 --"`                                      | Service responds with a JSON-RPC error; no panic or leaked stack trace       |
| 6 | Run concurrency test (100 parallel mixed calls) with `go test -race`| No data races; all responses pass schema and behavioural checks              |

This suite will lock down the JSON-RPC API’s **correctness, robustness, security, and contract compliance**, catching regressions early while remaining runnable on any developer workstation.

### LINKED ISSUE #726: Add unit tests for new JSON-RPC methods (PRT dispute entities)
## Context / Problem

We recently added new JSON-RPC methods for PRT dispute data, but the `internal/jsonrpc` package does not yet have unit tests covering them. Existing JSON-RPC methods are already being tested; only the new ones are missing coverage.

Methods to cover:
- `cartesi_getTournament`
- `cartesi_listTournaments`
- `cartesi_getCommitment`
- `cartesi_listCommitments`
- `cartesi_getMatch`
- `cartesi_listMatches`
- `cartesi_getMatchAdvanced`
- `cartesi_listMatchAdvances`

## Scope

Add unit tests in the `internal/jsonrpc` package to validate:
- request parsing (params decoding and validation)
- repository interaction (correct repo method called with expected arguments)
- response shape and error mapping (not found, invalid params, internal errors)

Out of scope:
- changing method semantics or response formats

## Deliverables

- [ ] Add unit tests for each method listed above.
- [ ] Add/extend repository mocks/fakes used by `internal/jsonrpc` tests to support tournaments/commitments/matches/match_advances queries.
- [ ] Ensure tests cover at least:
  - [ ] happy path (single get / list)
  - [ ] pagination path for list methods (when applicable)
  - [ ] invalid params (missing required param, malformed hex/uint, invalid address/name)
  - [ ] not found behavior for `get*` methods
  - [ ] repository error propagation mapped to JSON-RPC errors

## Acceptance criteria

- [ ] All listed methods have unit test coverage in `internal/jsonrpc` (or the package where JSON-RPC handlers live).
- [ ] Each `get*` method test includes:
  - [ ] success response
  - [ ] not found response
  - [ ] invalid params response
- [ ] Each `list*` method test includes:
  - [ ] success response
  - [ ] pagination arguments passed through correctly (limit/offset, or equivalent)
  - [ ] invalid params response
- [ ] Tests assert the JSON-RPC response structure (result vs error) and error codes/messages follow the project conventions.
- [ ] `make test` passes.

## Notes

- Prefer mirroring the style/patterns used by existing JSON-RPC handler tests in this repo (same helpers, same mock style, same error assertions).

### PR DESCRIPTION
**Pull request by mpolitzer**
closes https://github.com/cartesi/rollups-node/issues/682

### CODE REVIEW

File: internal/jsonrpc/jsonrpc.go [hunk 1]
```diff
@@ -1367,7 +1383,7 @@ func handleGetMatchAdvanced(s *Service, w http.ResponseWriter, r *http.Request,
 	}
 
 	matchAdvanced, err := s.repository.GetMatchAdvanced(r.Context(), params.Application, epochIndex,
-		params.TournamentAddress, params.IDHash, params.Parent)
+		params.TournamentAddress, params.IDHash, params.Parent[2:]) // TODO: use parsed value
```
[the remark this answers is not in the document]
Comment (mpolitzer) [reply internal/jsonrpc/jsonrpc.go:1367]: Moved to a separate issue: https://github.com/cartesi/rollups-node/issues/736

File: internal/jsonrpc/jsonrpc.go [hunk 2]
```diff
@@ -923,6 +923,10 @@ func handleListTournaments(s *Service, w http.ResponseWriter, r *http.Request, r
 		writeRPCError(w, req.ID, JSONRPC_INTERNAL_ERROR, "Internal server error", nil)
 		return
 	}
+	if len(tournaments) == 0 && !s.applicationExists(w, r, req, params.Application) {
+		writeRPCError(w, req.ID, JSONRPC_RESOURCE_NOT_FOUND, "Application not found", nil)
+		return
+	}
```
[the remark this answers is not in the document]
Comment (mpolitzer) [reply internal/jsonrpc/jsonrpc.go]: Refactored `applicationExists` function into `applicationAbsentOrError` for better semantic compression. Now the function wraps both possible `writeRPCError`s and the caller just has to `return` if something went wrong.

File: internal/jsonrpc/jsonrpc_test.go [hunk 3]
```diff
@@ -1536,18 +1538,1286 @@ func TestMethod(t *testing.T) {
...
+			body := s.doRequest(t, 0, fmt.Appendf([]byte{}, `{
+				"jsonrpc": "2.0",
+				"method": "cartesi_getCommitment",
+				"params": {
+				"application": "%v",
+				"epoch_index": "%v",
+				"tournament_address": "0x%020x"
+				},
+				"id": 0
+				}`, numberToName(app), hexutil.EncodeUint64(nr+1), 0))
```
[the remark this answers is not in the document]
Comment (mpolitzer) [reply internal/jsonrpc/jsonrpc_test.go]: Added a guard against empty commitment strings. And added `commitment` field to the test.

**Review by vfusco — DISMISSED** [review]

**Review by vfusco — APPROVED** [review]
